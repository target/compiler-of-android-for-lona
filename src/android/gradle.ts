export type AndroidConfig = {
  compileSdkVersion: number
  buildToolsVersion: string
  defaultConfig: {
    minSdkVersion: number
    targetSdkVersion: number
    versionCode: number
    versionName: string
  }
}

export type BuildScript = {
  plugins: string[]
  android: AndroidConfig
  dependencies: string[]
}

const createApplyPluginStatement = (plugin: string): string =>
  `apply plugin: "${plugin}"`

const createAndroidBlock = (android: AndroidConfig): string[] =>
  `android {
    compileSdkVersion ${android.compileSdkVersion}
    buildToolsVersion "${android.buildToolsVersion}"

    defaultConfig {
        minSdkVersion ${android.defaultConfig.minSdkVersion}
        targetSdkVersion ${android.defaultConfig.targetSdkVersion}
        versionCode ${android.defaultConfig.versionCode}
        versionName "${android.defaultConfig.versionName}"
    }
}`.split('\n')

const createDependencyStatement = (dependency: string): string =>
  `implementation "${dependency}"`

const createBlock = (identifier: string, statements: string[]): string[] =>
  statements.length === 0
    ? []
    : [
        `${identifier} {`,
        ...statements.map(statement => `    ${statement}`),
        `}`,
      ]

const addGroupSeparator = <Element, Group extends Array<Element>>(
  array: Group[],
  separator: Element
): Element[] => {
  return array
    .filter(inner => inner.length > 0)
    .flatMap((inner, index) =>
      index < array.length - 1 ? [...inner, separator] : inner
    )
}

export function createBuildScript(config: BuildScript): string {
  const applyPluginStatements = config.plugins.map(createApplyPluginStatement)
  const androidBlock = createAndroidBlock(config.android)
  const dependencyStatements = createBlock(
    'dependencies',
    config.dependencies.map(createDependencyStatement)
  )

  return addGroupSeparator(
    [applyPluginStatements, androidBlock, dependencyStatements],
    ''
  ).join('\n')
}

export const DEFAULT_PLUGINS = [
  'com.android.library',
  'kotlin-android',
  'kotlin-android-extensions',
]

export const DEFAULT_DEPENDENCIES = [
  'org.jetbrains.kotlin:kotlin-stdlib-jdk7:$kotlin_version',
  'androidx.appcompat:appcompat:1.1.0',
  'androidx.core:core-ktx:1.2.0',
]

export const DEFAULT_ANDROID_CONFIG: AndroidConfig = {
  compileSdkVersion: 29,
  buildToolsVersion: '29.0.2',
  defaultConfig: {
    minSdkVersion: 21,
    targetSdkVersion: 29,
    versionCode: 1,
    versionName: '1.0.0',
  },
}

export const DEFAULT_BUILD_CONFIG: BuildScript = {
  plugins: DEFAULT_PLUGINS,
  android: DEFAULT_ANDROID_CONFIG,
  dependencies: DEFAULT_DEPENDENCIES,
}
