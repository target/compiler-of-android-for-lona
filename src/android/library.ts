import path from 'path'
import { createFs, IFS, copy } from 'buffs'
import * as XML from '../xml/ast'
import { printElement } from '../xml/print'
import { formatDrawableName } from '../svg/drawable'
import { createKotlinAssetGallery } from './gallery'

/**
 * Create an Android manifest file for a library module.
 *
 * @param packageName Identifier like "com.lona.my_library"
 */
export function createManifest(packageName: string): string {
  const root: XML.Element = {
    tag: 'manifest',
    attributes: [
      {
        name: 'xmlns:android',
        value: 'http://schemas.android.com/apk/res/android',
      },
      {
        name: 'package',
        value: packageName,
      },
    ],
    content: [],
  }

  return printElement(root)
}

export type Library = {
  packageName: string
  androidManifest?: string
  buildScript?: string
  colorResources?: string
  elevationResources?: string
  textStyleResources?: string
  drawableResources: [string, IFS][]
}

/**
 * Generate an Android library module at the specified path.
 *
 * The following files may be generated:
 *
 * .
 * ├── .gitignore
 * ├── build.gradle
 * └── src
 *     └── main
 *         ├── AndroidManifest.xml
 *         ├── java
 *         │   └── ${packageName}
 *         │       └── gallery
 *         │           └── Gallery.kt
 *         └── res
 *             ├── drawable
 *             │   └── <various drawables>
 *             └── values
 *                 └── <various values>.xml
 *
 * @param rootPath
 * @param library
 * @returns An in-memory filesystem containing the generated files
 */
export function createLibraryFiles(rootPath: string, library: Library) {
  const target = createFs(
    {
      '.gitignore': '/build\n',
      ...(library.buildScript && { 'build.gradle': library.buildScript }),
      ...(library.androidManifest && {
        'src/main/AndroidManifest.xml': library.androidManifest,
      }),
      ...(library.colorResources && {
        'src/main/res/values/colors.xml': library.colorResources,
      }),
      ...(library.elevationResources && {
        'src/main/res/values/elevations.xml': library.elevationResources,
      }),
      ...(library.textStyleResources && {
        'src/main/res/values/text-styles.xml': library.textStyleResources,
      }),
      ...(library.drawableResources.length > 0 && {
        [path.join(
          'src/main/java',
          library.packageName.replace(/[.]/g, '/'),
          'gallery/Gallery.kt'
        )]: createKotlinAssetGallery(
          library.packageName,
          library.drawableResources.map(pair => formatDrawableName(pair[0]))
        ),
      }),
    },
    rootPath
  )

  library.drawableResources.forEach(([key, source]) => {
    copy(source, target.fs, '/', path.join(rootPath, 'src/main/res'))
  })

  return target
}
