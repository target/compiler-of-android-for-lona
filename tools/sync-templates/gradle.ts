import os from 'os'
import fs from 'fs-extra'
import path from 'path'
import { execSync } from 'child_process'

enum ProjectType {
  basic = 'basic',
}

enum DSL {
  groovy = 'groovy',
  kotlin = 'kotlin',
}

const commands = {
  gradleInit: ({
    type,
    dsl,
    projectName,
    projectDirectory,
  }: {
    type: ProjectType
    dsl: DSL
    projectName: string
    projectDirectory: string
  }): string => {
    return `gradle init --type ${type} --dsl ${dsl} --project-name ${projectName} --project-dir ${projectDirectory}`
  },
}

export default function sync() {
  // Create a temp directory
  const directoryPath = fs.mkdtempSync(path.join(os.tmpdir(), 'gradle-'))

  console.log(`Initializing gradle project in ${directoryPath}`)

  // Initialize a new gradle project
  execSync(
    commands.gradleInit({
      type: ProjectType.basic,
      dsl: DSL.groovy,
      projectName: 'temp-gradle-project',
      projectDirectory: directoryPath,
    })
  )

  const gradleWrapperTemplatePath = path.join(
    __dirname,
    '../../templates/android-studio/gradle/wrapper'
  )

  // Remove any old wrapper
  fs.removeSync(gradleWrapperTemplatePath)

  const wrapperFiles = ['gradle', 'gradlew', 'gradlew.bat']

  for (const file of wrapperFiles) {
    const sourcePath = path.join(directoryPath, file)
    const targetPath = path.join(gradleWrapperTemplatePath, file)

    console.log(`Copying gradle file ${sourcePath} to ${targetPath}`)

    // Copy new templates
    fs.copySync(sourcePath, targetPath)
  }

  // Clean up temp directory
  fs.removeSync(directoryPath)
}
