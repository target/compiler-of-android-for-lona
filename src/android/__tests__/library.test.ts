import { createBuildScript, DEFAULT_BUILD_CONFIG } from '../gradle'
import { createLibraryFiles, createManifest } from '../library'
import { createFs } from 'buffs'

describe('Android / Library', () => {
  test('creates a library module', () => {
    const { volume } = createLibraryFiles('/', {
      packageName: 'com.lona.test',
      buildScript: createBuildScript(DEFAULT_BUILD_CONFIG),
      androidManifest: createManifest('com.lona.test'),
      drawableResources: [
        ['test.svg', createFs({ 'drawable/test.xml': '' }, '/').fs],
      ],
    })

    expect(volume.toJSON()).toMatchSnapshot()
  })
})
