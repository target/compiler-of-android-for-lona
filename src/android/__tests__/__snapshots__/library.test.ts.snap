// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Android / Library creates a library module 1`] = `
Object {
  "/.gitignore": "/build
",
  "/build.gradle": "apply plugin: \\"com.android.library\\"
apply plugin: \\"kotlin-android\\"
apply plugin: \\"kotlin-android-extensions\\"

android {
    compileSdkVersion 29
    buildToolsVersion \\"29.0.2\\"

    defaultConfig {
        minSdkVersion 21
        targetSdkVersion 29
        versionCode 1
        versionName \\"1.0.0\\"
    }
}

dependencies {
    implementation \\"org.jetbrains.kotlin:kotlin-stdlib-jdk7:$kotlin_version\\"
    implementation \\"androidx.appcompat:appcompat:1.1.0\\"
    implementation \\"androidx.core:core-ktx:1.2.0\\"
}",
  "/src/main/AndroidManifest.xml": "<manifest
    xmlns:android=\\"http://schemas.android.com/apk/res/android\\"
    package=\\"com.lona.test\\" />",
  "/src/main/java/com/lona/test/gallery/Gallery.kt": "package com.lona.test.gallery

import com.lona.test.R

object Gallery {

    /*
     * A list of all generated assets.
     */
    val ALL_ASSETS = listOf(
        GalleryItem(\\"test\\", R.drawable.test)
    )

    /**
     * An item representing a generated asset.
     */
    data class GalleryItem(val id: String, val resourceId: Int)
}
",
  "/src/main/res/drawable/test.xml": "",
}
`;
